### PR TITLE
feat(ui): placeholders updated

### PR DESCRIPTION
**Pull request by neSpecc**
## Problem

1. Placeholders are being hidden on focus. It's not the best UX.
2. `autofocus` ruins `placeholder` option because of 1.
3. There are separated placeholder implementations in Tools CSS code.
4. Current implementation relies on CSS `:empty` state, but the is often a case when input has just `<br>` tag. Placeholder won't be shown in this case.

## Solution

1. Placeholder implementation moved to editor core. Now anyone can add `data-placeholder` to the contenteditable element inside editor and it will work.
2. Placeholders won't be hidden on focus
3. Now we're watching all inputs and mark them as empty via the `data-empty="true"` attribute if they are really empty.
4. We also support `data-placeholder-active` attribute to show placeholder only for current block. It is used for Paragraph tool — we don't want to show placeholders to each empty paragraphs on a page, only for the current.

Also, paragraph tool has a hack for showing `data-placeholder-active` event for not-current block in case if it is the only paragraph in empty editor. It is needed to make `placeholder` config option work when `autofocus` is false. See https://github.com/editor-js/paragraph/pull/89

https://github.com/codex-team/editor.js/assets/3684889/d9e09635-b09b-43c7-ab10-b1d9c2048a7e

### CODE REVIEW

File: src/components/block/index.ts [hunk 1]
```diff
@@ -262,6 +255,12 @@ export default class Block extends EventsDispatcher<BlockEvents> {
        * so we need to track focus events to update current input and clear cache.
        */
       this.addInputEvents();
+
+      /**
+       * We mark inputs with [data-empty] attribute
+       * It can be useful for developers, for example for correct placeholder behavior and
```
Comment (TatianaFomina) [inline src/components/block/index.ts]: and...?

File: src/components/modules/blockManager.ts [hunk 2]
```diff
@@ -873,7 +873,7 @@ export default class BlockManager extends Module {
    * Sets current Block Index -1 which means unknown
    * and clear highlights
    */
-  public dropPointer(): void {
+  public unsetCurrentBlock(): void {
```
Comment (TatianaFomina) [inline src/components/modules/blockManager.ts:876]: maybe `clearCurrentBlock()`?

Comment (neSpecc) [reply src/components/modules/blockManager.ts:876]: "clear" means that block will became empty, but we're not changing the block itself

**Review by TatianaFomina — APPROVED** [review]

File: test/cypress/tests/ui/DataEmpty.cy.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import { createEditorWithTextBlocks } from '../../support/utils/createEditorWithTextBlocks';
+
+describe('inputs [data-empty] mark', function () {
+  it('should be added to the editor on initialization', function () {
```
Comment (ilyamore88) [inline test/cypress/tests/ui/DataEmpty.cy.ts]:
```suggestion
  it('should be added to blocks on the editor initialization', function () {
```

Comment (ilyamore88) [reply test/cypress/tests/ui/DataEmpty.cy.ts]: The test case name sounds like this attribute should be added to the editor container.

File: test/cypress/tests/ui/DataEmpty.cy.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      .first()
+      .should('have.attr', 'data-empty', 'false');
+
+    cy.get('[data-cy=editorjs]')
+      .find('.ce-paragraph')
+      .last()
+      .should('have.attr', 'data-empty', 'true');
+  });
+
+  it('should be added to the block on input', function () {
```
Comment (ilyamore88) [inline test/cypress/tests/ui/DataEmpty.cy.ts]: Your test case checks that this attr should be toggled on input. Would be great to add a test case that this attr will be set to true if this block is empty on input.

File: test/cypress/tests/ui/DataEmpty.cy.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      .last()
+      .type('Some text');
+
+    cy.get('[data-cy=editorjs]')
+      .find('.ce-paragraph')
+      .last()
+      .should('have.attr', 'data-empty', 'false');
+  });
+
+  it('should be added to the block on focus', function () {
```
Comment (ilyamore88) [inline test/cypress/tests/ui/DataEmpty.cy.ts]: It's good to cover this case, but if this behaviour is broken, this case will pass because of adding this attr by default on initialisation. I don't know how to test it in the right way.

File: test/cypress/tests/ui/DataEmpty.cy.ts (hunk 3 above)
Comment (neSpecc) [reply test/cypress/tests/ui/DataEmpty.cy.ts]: fixed

File: test/cypress/tests/ui/DataEmpty.cy.ts (hunk 4 above)
Comment (neSpecc) [reply test/cypress/tests/ui/DataEmpty.cy.ts]: added `should be added as "true" to the input on chars removal`

File: test/cypress/tests/ui/DataEmpty.cy.ts (hunk 5 above)
Comment (neSpecc) [reply test/cypress/tests/ui/DataEmpty.cy.ts]: removed this test

File: test/cypress/tests/ui/Placeholders.cy.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    cy.createEditor({
+      placeholder: PLACEHOLDER_TEXT,
+      autofocus: true,
+    });
+
+    cy.get('[data-cy=editorjs]')
+      .find('.ce-paragraph')
+      .getPseudoElementContent('::before')
+      .should('eq', PLACEHOLDER_TEXT);
+  });
```
Comment (ilyamore88) [inline test/cypress/tests/ui/Placeholders.cy.ts:36]: So in both cases when I pass a placeholder, it will be shown. It doesn't really depend on the autofocus param. The second case is redundant then.

File: test/cypress/tests/ui/Placeholders.cy.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  it('should be shown event if input is focused', function () {
```
Comment (ilyamore88) [inline test/cypress/tests/ui/Placeholders.cy.ts:38]:
```suggestion
  it('should be shown even if input is focused', function () {
```

File: test/cypress/tests/ui/Placeholders.cy.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    cy.get('[data-cy=editorjs]')
+      .find('.ce-paragraph')
+      .click()
+      .as('firstBlock')
+      .getPseudoElementContent('::before')
+      .should('eq', PLACEHOLDER_TEXT);
+  });
+
+  it('should be shown event when user removes all text by cmd+a and delete', function () {
```
Comment (ilyamore88) [inline test/cypress/tests/ui/Placeholders.cy.ts:51]:
```suggestion
  it('should be shown even when user removes all text by cmd+a and delete', function () {
```

Comment (ilyamore88) [reply test/cypress/tests/ui/Placeholders.cy.ts:51]: Should be shown in an empty element when a user removes all content.

File: test/cypress/tests/ui/Placeholders.cy.ts (hunk 6 above)
Comment (neSpecc) [reply test/cypress/tests/ui/Placeholders.cy.ts:36]: in previous version `autofocus` was leading placeholder to disappear, so I've added this test.
